pkg/endpoint/endpoint.go

```diff
@@ -299,6 +300,12 @@ type Endpoint struct {
 
 	EventQueue *eventqueue.EventQueue `json:"-"`
 
+	// DatapathConfiguration is the endpoint's datapath configuration as
+	// passed in via the plugin that created the endpoint, e.g. the CNI
+	// plugin which performed the plumbing will enable certain datapath
+	// features according to the mode selected.
+	DatapathConfiguration models.EndpointDatapathConfiguration
```

Comment:
Maybe I missed this, but how is upgrade handled with respect to this configuration?

Reply:
Right now, there are no plans to support upgrade. I've tried it and it is possible. Some of the required work was previously in this branch but this PR has already gotten big so I've removed it. As it stands now, whether to use per endpoint routes and direct linux routing is a decision made at cluster creation time, similar to IPVLAN or flannel mode.

Details:
Upgrade is possible but it requires new programs to keep a policy enforcement at egress as old endpoints may not protect themselves yet. I have a commit which does this via a flag that is enabled by default but disabled in new ConfigMaps. We could also consider automatically ripping that out once all old endpoints are gone. For now, I have chosen to reserve this discussion for a separate PR as it is not strictly required to establish the functionality of ENI routing and additional chaining.

Reply:
I wasn't necessarily thinking of upgrade in terms of transitioning existing setups to the new model, just in terms of "is this populated correctly to achieve the old model when going from an existing setup" (minimally, try not to break things). I'm happy to push this out.

Reply:
Oh, I see. All fields are opt-in except RequireRouting which defaults to true. The `RequireRouting()` on `Endpoint` will default to true when the field is not available.